x/pki/types/message_add_pki_revocation_distribution_point.go

```diff
@@ -0,0 +1,135 @@
+
+			break
+		}
+	}
+
+	if !isDataDigestInTypes {
+		return pkitypes.NewErrInvalidDataDigestType(fmt.Sprintf("invalid DataDigestType: %d", msg.DataDigestType))
+	}
+
+	if msg.RevocationType != allowedRevocationType {
+		return pkitypes.NewErrInvalidRevocationType(fmt.Sprintf("invalid RevocationType: %d", msg.RevocationType))
+	}
+
+	cert, err := x509.DecodeX509Certificate(msg.CrlSignerCertificate)
+	if err != nil {
+		return err
+	}
+
+	if msg.IsPAA {
+		if msg.Pid != 0 {
+			return pkitypes.NewErrNotEmptyPid("pid is not empty for root certificate")
+		}
+
+		if !cert.IsSelfSigned() {
+			return pkitypes.NewErrPAANotSelfSigned(fmt.Sprintf("root certificate with subject: %s, subjectKeyID: %s is not self-signed", cert.SubjectAsText, cert.SubjectKeyID))
+		}
+	} else {
+		if !strings.Contains(cert.SubjectAsText, string(msg.Pid)) {
+			return pkitypes.NewErrCRLSignerCertificateDoesNotContainPid(fmt.Sprintf("CRLSignerCertificate with subject: %s, subjectKeyID does not contain pid: %d", cert.SubjectAsText, cert.SubjectKeyID, msg.Pid))
+		}
+
+		if cert.IsSelfSigned() {
+			return pkitypes.NewErrNonPAASelfSigned(fmt.Sprintf("non-root certificate with subject: %s, subjectKeyID: %s is self-signed", cert.SubjectAsText, cert.SubjectKeyID))
+		}
+	}
+
+	if msg.DataFileSize == 0 && msg.DataDigest != "" {
+		return pkitypes.NewErrEmptyDataFileSize(fmt.Sprintf("msgAddRevocationDistributionPoint with CRLSignerCertificate: %s has empty DataFileSize when DataDigest is not empty", msg.CrlSignerCertificate))
+	}
+
+	if msg.DataFileSize != 0 && msg.DataDigest == "" {
+		return pkitypes.NewErrEmptyDataDigest(fmt.Sprintf("msgAddRevocationDistributionPoint with CRLSignerCertificate: %s has empty DataDigest when DataFileSize is not empty", msg.CrlSignerCertificate))
+	}
+
+	if msg.DataDigest == "" && msg.DataDigestType != 0 {
+		return pkitypes.NewErrEmptyDataDigest(fmt.Sprintf("msgAddRevocationDistributionPoint with CRLSignerCertificate: %s has empty DataDigest when DataDigestType is not empty", msg.CrlSignerCertificate))
+	}
+
+	if msg.DataDigest != "" && msg.DataDigestType == 0 {
+		return pkitypes.NewErrEmptyDataDigestType(fmt.Sprintf("msgAddRevocationDistributionPoint with CRLSignerCertificate: %s has empty DataDigestType when DataDigest is not empty", msg.CrlSignerCertificate))
+	}
+
+	if msg.RevocationType == 1 && (msg.DataFileSize != 0 || msg.DataDigest != "" || msg.DataDigestType != 0) {
+		return pkitypes.NewErrDataFieldPresented(fmt.Sprintf("msgAddRevocationDistributionPoint with CRLSignerCertificate: %s has one or more non-empty DataFields when RevocationType is 1", msg.CrlSignerCertificate))
+	}
+
+	if msg.IssuerSubjectKeyID != cert.SubjectKeyID {
+		return pkitypes.NewErrIssuerSubjectKeyIDNotEqualsCertSubjectKeyID(fmt.Sprintf("msgAddRevocationDistributionPoint with CRLSignerCertificate: %s has IssuerSubjectKeyID: %s which is not equal to certificate SubjectKeyID: %s", msg.CrlSignerCertificate, msg.IssuerSubjectKeyID, cert.SubjectKeyID))
+	}
+
```

Comment:
The following checks are missing:
1) `IssuerSubjectKeyID` must consist of even number of uppercase hexadecimal characters ([0-9A-F]), with no whitespace and no non-hexadecimal characters., e.g: `5A880E6C3653D07FB08971A3F473790930E62BDB`.
2) `DataUrl` starts with either `http` or `https`

Reply:
1. Fixed
2. Checked in validator.Validate()